topdown/subset.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2022 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+func bothObjects(t1, t2 *ast.Term) (bool, ast.Object, ast.Object) {
+	if (t1 == nil) || (t2 == nil) {
+		return false, nil, nil
+	}
+
+	obj1, ok := t1.Value.(ast.Object)
+	if !ok {
+		return false, nil, nil
+	}
+
+	obj2, ok := t2.Value.(ast.Object)
+	if !ok {
+		return false, nil, nil
+	}
+
+	return true, obj1, obj2
+}
+
+func bothSets(t1, t2 *ast.Term) (bool, ast.Set, ast.Set) {
+	if (t1 == nil) || (t2 == nil) {
+		return false, nil, nil
+	}
+
+	set1, ok := t1.Value.(ast.Set)
+	if !ok {
+		return false, nil, nil
+	}
+
+	set2, ok := t2.Value.(ast.Set)
+	if !ok {
+		return false, nil, nil
+	}
+
+	return true, set1, set2
+}
+
+func bothArrays(t1, t2 *ast.Term) (bool, *ast.Array, *ast.Array) {
+	if (t1 == nil) || (t2 == nil) {
+		return false, nil, nil
+	}
+
+	array1, ok := t1.Value.(*ast.Array)
+	if !ok {
+		return false, nil, nil
+	}
+
+	array2, ok := t2.Value.(*ast.Array)
+	if !ok {
+		return false, nil, nil
+	}
+
+	return true, array1, array2
```

Comment:
[ignore] Just for code golf purposes, this could be
```suggestion
	array1, ok1 := t1.Value.(*ast.Array)
	array2, ok2 := t2.Value.(*ast.Array)

	return ok1 && ok2, array1, array2
```

But it's not much clearer, so let's keep it as it is.

Reply:
Also, this is worse from a performance perspective, because you are paying the unboxing cost for both arrays up front. In the case where `t1` fails to unbox, the current approach lets us skip trying to unbox `t2`.

Reply:
Admittedly, the performance difference is probably extremely marginal. According to this [nice person on the internet](https://stackoverflow.com/a/28027945), type assertions apparently aren't expensive like they used to be in older Go versions.

Reply:
> they used to be in older Go versions

Oh they had been at some point? Kind of always assumed it was as good as free 😅